src/factories/calorimetry/CalorimeterClusterRecoCoG_factory.h

```diff
@@ -0,0 +1,62 @@
+// Copyright 2023, Wouter Deconinck
+// Subject to the terms in the LICENSE file found in the top-level directory.
+//
+
+#pragma once
+
+#include <algorithms/calorimetry/CalorimeterClusterRecoCoG.h>
+#include <services/geometry/dd4hep/JDD4hep_service.h>
+#include <extensions/jana/JOmniFactory.h>
+
+
+namespace eicrecon {
+
+class CalorimeterClusterRecoCoG_factory : public JOmniFactory<CalorimeterClusterRecoCoGConfig> {
+
+    const dd4hep::Detector* m_detector;
+    eicrecon::CalorimeterClusterRecoCoG m_algo;
+
+public:
+    void Init() override {
+
+        DeclarePodioOutput<edm4eic::Cluster>(GetOutputTags()[0]);
+        DeclarePodioOutput<edm4eic::MCRecoClusterParticleAssociation>(GetOutputTags()[1]);
+
+        DeclareParameter("samplingFraction", GetConfig().sampFrac);
+        DeclareParameter("logWeightBase", GetConfig().logWeightBase);
+        DeclareParameter("depthCorrection", GetConfig().depthCorrection);
+        DeclareParameter("energyWeight", GetConfig().energyWeight);
+        DeclareParameter("moduleDimZName", GetConfig().moduleDimZName);
+        DeclareParameter("enableEtaBounds", GetConfig().enableEtaBounds);
+
+        auto app = GetApplication();
+
+        // Use JDD4hep_service to get dd4hep::Detector
+        auto geoSvc = app->template GetService<JDD4hep_service>();
+        m_detector = geoSvc->detector();
+
+        m_algo.applyConfig(GetConfig());
+        m_algo.init(m_detector, logger());
+    }
+
+    //------------------------------------------
+    // Process
+    void Process(const std::shared_ptr<const JEvent> &event) override {
+
+        // TODO: NWB: We are using GetCollectionBase because GetCollection is temporarily out of commission due to JFactoryPodioTFixed
+        auto proto = static_cast<const edm4eic::ProtoClusterCollection*>(event->GetCollectionBase(GetInputTags()[0]));
+        auto mchits = static_cast<const edm4hep::SimCalorimeterHitCollection*>(event->GetCollectionBase(GetInputTags()[1]));
+
+        try {
+            auto clusters_with_assocs = m_algo.process(proto, mchits);
+            SetCollection<edm4eic::Cluster>(GetOutputTags()[0], std::move(clusters_with_assocs.first));
```

Comment:
At some point a generic algorithms approach will need to get collections and set collections for an arbitrary number of inputs and outputs, still by string tag. I was working on modifying `DeclarePodioOutputs` and `SetCollection` that allows a syntax like:
```suggestion
            SetCollections<std::tuple<edm4eic::Cluster, std::emd4eicMCRecoClusterParticleAssociation>>(GetOutputTags(), std::move(clusters_with_assocs));
```
But I got somewhat lost in `std::indexed_sequence`s and fold expressions needed for this :-)

Reply:
I know what you mean. I think that the variadic SetCollections() should live at the JOmniFactory level and not the JMultifactory level though. Once we do this, the only logic left in the factory is associating ConfigT fields with parameter names, and acquiring Services. Services are easy to handle via another parameter pack, and maybe the name associations should be attached to the ConfigT struct? However I want to get the story straight on how we wire everything from a config file before we go down another rabbit hole.

Reply:
I agree, this wasn't going into JMultiFactory but indeed needs the Omni factory (I had a different name in my experimentation).